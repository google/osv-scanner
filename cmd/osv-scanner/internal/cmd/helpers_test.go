package cmd

import (
	"bytes"
	"log/slog"
	"reflect"
	"testing"

	"github.com/google/osv-scanner/v2/internal/cmdlogger"
	"github.com/google/osv-scanner/v2/internal/testutility"
	"github.com/urfave/cli/v2"
)

func Test_insertDefaultCommand(t *testing.T) {
	commands := []*cli.Command{
		{Name: "default"},
		{Name: "helpers.go"},
		{Name: "scan"},
	}
	defaultCommand := "default"

	tests := []struct {
		originalArgs []string
		wantArgs     []string
	}{
		// test when default command is specified
		{
			originalArgs: []string{"", "default", "file"},
			wantArgs:     []string{"", "default", "source", "file"},
		},
		// test when command is not specified
		{
			originalArgs: []string{"", "file"},
			wantArgs:     []string{"", "default", "source", "file"},
		},
		// test when command is also a filename
		{
			originalArgs: []string{"", "helpers.go"},
			wantArgs:     []string{"", "helpers.go"},
		},
		// test when subcommand is also a filename
		{
			originalArgs: []string{"", "default", "image"},
			wantArgs:     []string{"", "default", "image"},
		},
		// test when command is not valid
		{
			originalArgs: []string{"", "invalid"},
			wantArgs:     []string{"", "default", "source", "invalid"},
		},
		// test when command is a built-in option
		{
			originalArgs: []string{"", "--version"},
			wantArgs:     []string{"", "--version"},
		},
		{
			originalArgs: []string{"", "-h"},
			wantArgs:     []string{"", "-h"},
		},
		{
			originalArgs: []string{"", "help"},
			wantArgs:     []string{"", "help"},
		},
	}

	for _, tt := range tests {
		stdout := &bytes.Buffer{}
		stderr := &bytes.Buffer{}

		logger := cmdlogger.New(stdout, stderr)

		slog.SetDefault(slog.New(&logger))

		argsActual := insertDefaultCommand(tt.originalArgs, commands, defaultCommand, stderr)
		if !reflect.DeepEqual(argsActual, tt.wantArgs) {
			t.Errorf("Test Failed. Details:\n"+
				"Args (Got):  %s\n"+
				"Args (Want): %s\n", argsActual, tt.wantArgs)
		}
		testutility.NewSnapshot().MatchText(t, stdout.String())
		testutility.NewSnapshot().MatchText(t, stderr.String())
	}
}
