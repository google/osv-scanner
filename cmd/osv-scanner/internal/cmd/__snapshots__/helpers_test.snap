
[Test_insertDefaultCommand - 1]

---

[Test_insertDefaultCommand - 2]

---

[Test_insertDefaultCommand - 3]

---

[Test_insertDefaultCommand - 4]

---

[Test_insertDefaultCommand - 5]

---

[Test_insertDefaultCommand - 6]
Warning: `helpers.go` exists as both a subcommand of OSV-Scanner and as a file on the filesystem. `helpers.go` is assumed to be a subcommand here. If you intended for `helpers.go` to be an argument to `default`, you must specify `default helpers.go` in your command line.

---

[Test_insertDefaultCommand - 7]

---

[Test_insertDefaultCommand - 8]

---

[Test_insertDefaultCommand - 9]

---

[Test_insertDefaultCommand - 10]

---

[Test_insertDefaultCommand - 11]

---

[Test_insertDefaultCommand - 12]

---

[Test_insertDefaultCommand - 13]

---

[Test_insertDefaultCommand - 14]

---

[Test_insertDefaultCommand - 15]

---

[Test_insertDefaultCommand - 16]

---
